src/lib/cache.coffee

```diff
@@ -0,0 +1,83 @@
+###
+
+dnschain
+http://dnschain.net
+
+Copyright (c) 2014 okTurtles Foundation
+
+This Source Code Form is subject to the terms of the Mozilla Public
+License, v. 2.0. If a copy of the MPL was not distributed with this
+file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+###
+
+redis = require 'redis'
+
+module.exports = (dnschain) ->
+    # expose these into our namespace
+    for k of dnschain.globals
+        eval "var #{k} = dnschain.globals.#{k};"
+
+    class ResolverCache
+        constructor: (@dnschain) ->
+            @log = gNewLogger 'CACHE'
+            @log.info "Loading Redis Cache..."
+            if gConf.get('redis:blockchain:enabled') or gConf.get('redis:oldDNS:enabled')
+                @log.info "cache enabled"
+                [host,port] = gConf.get('redis:socket').split(':')
+                @cache = if port?
+                    port = parseInt port
+                    redis.createClient(port, host)
+                else
+                    redis.createClient(host)
+                if gConf.get('redis:blockchain:enabled')
+                    @blockchainEnabled = true
+                if gConf.get('redis:oldDNS:enabled')
+                    @oldDNSEnabled = true
+                    @oldDNSTTL = 600
+                @cache.on 'error', (err) =>
+                    @log.error "cache errored"
+                    @shutdown()
+
+            else
+                @log.info "cache disabled"
+
+        resolveBlockchain: (resolver, path, options, cb) ->
+            if @blockchainEnabled? and resolver.cacheTTL?
+                @cache.get "#{resolver.name}:#{path}:#{JSON.stringify(options)}", (err, result) =>
```

Comment:
Giving this more though, I personally thing (2) is the better choice.

After all, we don't know how many more uses we will have for Redis, and by making a specific method for each action you are in fact repeating yourself.

In principle, Redis should just be viewed as an abstract key/value store (that's what it is, after all) that takes special actions when a value is not found. This can be accomplished very easily using generic, simple `get:` and `set:` methods.

For those methods, I think you should be able to specify a callback that gets called (as needed) if, for example, a value is not found. Then the keys that are being looked up will be defined outside of this class.

Doable?